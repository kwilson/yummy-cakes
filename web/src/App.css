.app {
  display: flex;
  flex-direction: column;
  height: 100%;
}

.app > * {
  padding: 1em 2em;
}

.app__header {
  background: #efefef;
}

.app__body {
  flex: 1;
  overflow: auto;
}

.app__footer {
  background-color:#333;
  color: #eee;
  text-align: right;
}

