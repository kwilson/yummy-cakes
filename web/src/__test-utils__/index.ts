export function asMock<T>(fn: (...params: any[]) => T): jest.Mock<T> {
  expect(jest.isMockFunction(fn)).toBe(true);
  return fn as jest.Mock<T>;
}
