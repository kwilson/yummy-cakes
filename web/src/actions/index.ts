import { ThunkAction } from 'redux-thunk';
import rest from '../reducers/rest';
import { isCakesListLoading } from '../selectors';
import { AppState } from '../reducers';
import { NewCakeModel } from '../models/NewCakeModel';

export const SELECT_CAKE = 'SELECT_CAKE';
export const CLEAR_SELECTED_CAKE = 'CLEAR_SELECTED_CAKE';

export const SHOW_SUBMIT_CAKE = 'SHOW_SUBMIT_CAKE';
export const CANCEL_SUBMIT_CAKE = 'CANCEL_SUBMIT_CAKE';

const fetchCakes = (): ThunkAction<void, AppState, void> => dispatch => {
  return dispatch(rest.actions.cakes());
};

const shouldFetchCakes = (state: AppState) => {
  return !isCakesListLoading(state);
};

export const selectCake = (id: string) => ({
  type: SELECT_CAKE,
  payload: {
    id
  }
});

export const clearSelectedCake = () => ({
  type: CLEAR_SELECTED_CAKE
});

export const fetchCakesIfNeeded = (): ThunkAction<void, AppState, void> => (dispatch, getState) => {
  if (shouldFetchCakes(getState())) {
    return dispatch(fetchCakes());
  }
};

export const showSubmitCakeForm = () => ({
  type: SHOW_SUBMIT_CAKE
});

export const submitCake = (cake: NewCakeModel): ThunkAction<void, AppState, void> => (dispatch, getState) => {
  dispatch(
    rest.actions.cake(
      undefined, {
        body: JSON.stringify(cake)
      }
    ))
    .then(() => {
      dispatch(cancelSubmitCake());
      dispatch(fetchCakes());
    });
};

export const cancelSubmitCake = () => ({
  type: CANCEL_SUBMIT_CAKE
});
