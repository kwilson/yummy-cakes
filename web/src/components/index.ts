export { default as Cake } from './Cake/Cake';
export { default as CakeDetail } from './CakeDetail/CakeDetail';
export { default as CakeList } from './CakeList/CakeList';
export { default as SubmitCake } from './SubmitCake/SubmitCake';
export { default as YumFactor } from './YumFactor/YumFactor';
