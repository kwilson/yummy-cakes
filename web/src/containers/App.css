.app {
  display: flex;
  flex-direction: column;
  height: 100%;
}

.app > * {
  padding: 10px 15px;
}

.app__header {
  background: #efefef;
  display: flex;
  justify-content: space-between;
  align-items: baseline;
}

  .btn-show-submit-form {
    background: rgb(11, 78, 202);
    border: solid 1px white;
    color: white;
    cursor: pointer;
    padding: 8px 12px;
    font-size: 1.4em;
  }

.app__body {
  flex: 1;
  overflow: auto;
}

.app__footer {
  background-color:#333;
  color: #eee;
  text-align: right;
}

