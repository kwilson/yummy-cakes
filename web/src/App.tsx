import * as React from 'react';
import './App.css';

class App extends React.Component {
  render() {
    return (
      <div className="app">
        <header className="app__header">
          <h1>Yummy Cakes</h1>
        </header>

        <div className="app__body">
          <p>Body goes here</p>
        </div>

        <footer className="app__footer">Created by Kevin Wilson</footer>
      </div>
    );
  }
}

export default App;
