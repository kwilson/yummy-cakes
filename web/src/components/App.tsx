import * as React from 'react';
import { connect, Dispatch } from 'react-redux';
import './App.css';

import AllCakesList from '../containers/AllCakesList';
import { fetchCakesIfNeeded } from '../actions';
import { CakeModel } from '../models/CakeModel';

export interface AppState {
  restReducers: {
    cakes: {
      loading: boolean;
      data: {
        data?: CakeModel[]
      }
    }
  };
}

class App extends React.Component<{ dispatch: Dispatch<AppState> }> {
  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(fetchCakesIfNeeded());
  }

  render() {
    return (
      <div className="app">
        <header className="app__header">
          <h1>Yummy Cakes</h1>
        </header>

        <div className="app__body">
          <AllCakesList />
        </div>

        <footer className="app__footer">Created by Kevin Wilson</footer>
      </div>
    );
  }
}

export default connect()(App);
