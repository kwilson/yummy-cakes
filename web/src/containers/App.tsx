import * as React from 'react';
import { connect, Dispatch, MapStateToProps, MapDispatchToProps } from 'react-redux';
import './App.css';

import { fetchCakesIfNeeded, selectCake, clearSelectedCake, submitCake, cancelSubmitCake, showSubmitCakeForm } from '../actions';
import { AppState } from '../reducers';
import { getAllCakesList, isCakesListLoading, getSelectedCake, isSubmitFormVisible } from '../selectors';

import { CakeList, CakeDetail, SubmitCake } from '../components/';

import { CakeModel } from '../models/CakeModel';
import { NewCakeModel } from '../models/NewCakeModel';

export interface AppStateProps {
  allCakes: CakeModel[];
  selectedCake?: CakeModel;
  isSubmitFormVisible: boolean;
  isLoading: boolean;
}

export interface AppDispatchProps {
  dispatch: Dispatch<{}>;
  selectCake: (id: string) => any;
  clearSelectedCake: () => any;

  showSubmitCakeForm: () => any;
  submitCake: (cake: NewCakeModel) => any;
  cancelSubmitCake: () => any;
}

export const mapStateToProps: MapStateToProps<AppStateProps, void, AppState> = (state) => {
  return {
    allCakes: getAllCakesList(state),
    selectedCake: getSelectedCake(state),
    isSubmitFormVisible: isSubmitFormVisible(state),

    isLoading: isCakesListLoading(state)
  };
};

export const mapDispatchToProps: MapDispatchToProps<AppDispatchProps, void> = dispatch => {
  return {
    dispatch,
    selectCake: (id: string) => dispatch(selectCake(id)),
    clearSelectedCake: () => dispatch(clearSelectedCake()),

    showSubmitCakeForm: () => dispatch(showSubmitCakeForm()),
    submitCake: (cake: CakeModel) => dispatch(submitCake(cake)),
    cancelSubmitCake: () => dispatch(cancelSubmitCake())
  };
};

export class App extends React.Component<AppStateProps & AppDispatchProps> {
  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(fetchCakesIfNeeded());
  }

  render() {
    const onSelectCake = (id: string) => this.props.selectCake(id);
    const handleShowSubmitCakeForm = () => this.props.showSubmitCakeForm();

    return (
      <div className="app">
        <header className="app__header">
          <h1>Yummy Cakes</h1>
          <button className="btn-show-submit-form" onClick={handleShowSubmitCakeForm}>Submit A Cake</button>
        </header>

        <div className="app__body">
          <CakeList
            cakes={this.props.allCakes}
            onSelectCake={onSelectCake}
          />
          <CakeDetail
            selectedCake={this.props.selectedCake}
            close={this.props.clearSelectedCake}
          />

          {this.props.isSubmitFormVisible &&
            <SubmitCake
              submitCake={this.props.submitCake}
              cancel={this.props.cancelSubmitCake}
            />
          }
        </div>

        <footer className="app__footer">Created by Kevin Wilson</footer>
      </div>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);
