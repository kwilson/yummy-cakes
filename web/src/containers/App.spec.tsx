import * as React from 'react';
import { shallow } from 'enzyme';
import { create as createSnapshot } from 'react-test-renderer';

jest.mock('../selectors');
import { getAllCakesList, isCakesListLoading, getSelectedCake } from '../selectors';

jest.mock('../actions');
import { fetchCakesIfNeeded, selectCake, clearSelectedCake } from '../actions';

import { asMock } from '../__test-utils__';
import { CakeModel } from '../models/CakeModel';

import { App, mapStateToProps, mapDispatchToProps, AppDispatchProps } from './App';
import { AppState } from '../reducers';
import { CakeList, CakeDetail } from '../components';

describe('App Container', () => {
  describe('mapStateToProps', () => {
    it('maps values from selectors', () => {
      // Arrange
      const allCakes: CakeModel[] = [];
      const selectedCake: CakeModel = { id: 'id123' };
      const isLoading = true;

      const mockState = {} as AppState;

      asMock(getAllCakesList).mockReturnValue(allCakes);
      asMock(getSelectedCake).mockReturnValue(selectedCake);
      asMock(isCakesListLoading).mockReturnValue(isLoading);

      // Act
      const result = mapStateToProps(mockState, undefined);

      // Assert
      expect(getAllCakesList).toHaveBeenCalledWith(mockState);
      expect(getSelectedCake).toHaveBeenCalledWith(mockState);
      expect(isCakesListLoading).toHaveBeenCalledWith(mockState);

      expect(result.allCakes).toBe(allCakes);
      expect(result.selectedCake).toBe(selectedCake);
      expect(result.isLoading).toBe(isLoading);
    });
  });

  describe('mapDispatchToProps', () => {
    it('attaches to dispatch events', () => {
      // Arrange
      const mockDispatch = jest.fn();

      // Act
      // Need to cast to 'any' to get around issue with the
      // Typescript defs.
      // See: https://github.com/Microsoft/TypeScript/issues/14091
      const result: AppDispatchProps = (mapDispatchToProps as any)(mockDispatch);

      // Assert
      expect(result.dispatch).toBe(mockDispatch);

      // selectCake
      const selectCakeValue = {};
      asMock(selectCake).mockReturnValue(selectCakeValue);
      result.selectCake(expect.any(typeof Number));
      expect(mockDispatch).toHaveBeenCalledWith(selectCakeValue);

      // selectCake
      const clearSelectedCakeValue = {};
      asMock(clearSelectedCake).mockReturnValue(clearSelectedCakeValue);
      result.clearSelectedCake();
      expect(mockDispatch).toHaveBeenCalledWith(clearSelectedCakeValue);
    });
  });

  describe('App component', () => {
    const mockDispatch = jest.fn();
    const mockSelectCake = jest.fn();
    const mockClearSelectedCake = jest.fn();
    const mockSubmitCake = jest.fn();
    const mockCancelSubmitCake = jest.fn();
    const mockShowSubmitCakeForm = jest.fn();

    const allCakes: CakeModel[] = [];
    const isLoading = false;
    const selectedCake: CakeModel = { id: 'id123' };
    const isSubmitFormVisible = false;

    const fetchCakesIfNeededValue = {};

    afterEach(() => {
      mockDispatch.mockReset();
      mockSelectCake.mockReset();
      mockClearSelectedCake.mockReset();
      mockSubmitCake.mockReset();
      mockCancelSubmitCake.mockReset();
      mockShowSubmitCakeForm.mockReset();
    });

    it('dispatches fetchCakesIfNeeded on mount', () => {
      // Arrange
      asMock(fetchCakesIfNeeded).mockReturnValue(fetchCakesIfNeededValue);

      const cmp = (
        <App
          allCakes={allCakes}
          isSubmitFormVisible={isSubmitFormVisible}
          isLoading={isLoading}

          dispatch={mockDispatch}
          selectCake={mockSelectCake}
          clearSelectedCake={mockClearSelectedCake}
          submitCake={mockSubmitCake}
          cancelSubmitCake={mockCancelSubmitCake}
          showSubmitCakeForm={mockShowSubmitCakeForm}
        />
      );

      // Act
      shallow(cmp);
      const snapshot = createSnapshot(cmp).toJSON();

      // Assert
      expect(fetchCakesIfNeeded).toHaveBeenCalled();
      expect(mockDispatch).toHaveBeenCalledWith(fetchCakesIfNeededValue);

      expect(snapshot).toMatchSnapshot();
    });

    it('passes props to child components', () => {
      // Arrange
      const cmp = (
        <App
          allCakes={allCakes}
          isSubmitFormVisible={isSubmitFormVisible}
          isLoading={isLoading}

          dispatch={mockDispatch}
          selectCake={mockSelectCake}
          clearSelectedCake={mockClearSelectedCake}
          submitCake={mockSubmitCake}
          cancelSubmitCake={mockCancelSubmitCake}
          showSubmitCakeForm={mockShowSubmitCakeForm}
        />
      );

      // Act
      const el = shallow(cmp);
      const snapshot = createSnapshot(cmp).toJSON();

      // Assert
      expect(el.find(CakeList).props().cakes).toBe(allCakes);
      el.find(CakeList).props().onSelectCake('test');
      expect(selectCake).toHaveBeenCalledWith('test');

      expect(el.find(CakeDetail).props().selectedCake).toBe(selectedCake);
      el.find(CakeDetail).props().close();
      expect(clearSelectedCake).toHaveBeenCalled();

      expect(snapshot).toMatchSnapshot();
    });
  });
});
