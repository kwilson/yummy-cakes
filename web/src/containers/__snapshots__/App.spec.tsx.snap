// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Container App component dispatches fetchCakesIfNeeded on mount 1`] = `
<div
  className="app"
>
  <header
    className="app__header"
  >
    <h1>
      Yummy Cakes
    </h1>
  </header>
  <div
    className="app__body"
  >
    <div
      className="cake-list--empty"
    >
      No cakes to view
    </div>
  </div>
  <footer
    className="app__footer"
  >
    Created by Kevin Wilson
  </footer>
</div>
`;

exports[`App Container App component passes props to child components 1`] = `
<div
  className="app"
>
  <header
    className="app__header"
  >
    <h1>
      Yummy Cakes
    </h1>
  </header>
  <div
    className="app__body"
  >
    <div
      className="cake-list--empty"
    >
      No cakes to view
    </div>
    <div
      className="cake-detail"
    >
      <div
        className="cake-detail__img-container"
      >
        <img
          className="cake-detail__img"
          src={undefined}
        />
      </div>
      <h1
        className="cake-detail__heading"
      />
      <div
        className="cake-detail__yum"
      />
      <div
        className="cake-detail__comment"
      >
        <h2>
          Comments
        </h2>
        <div
          className="cake-detail__comment-value"
        >
          No comments
        </div>
      </div>
      <button
        className="cake-detail__close"
        onClick={[Function]}
      >
        Close
      </button>
    </div>
  </div>
  <footer
    className="app__footer"
  >
    Created by Kevin Wilson
  </footer>
</div>
`;
